examples/react-dapp/src/App.tsx

```diff
@@ -18,52 +21,129 @@ type Role = {
   namespace: string
 }
 
-type User = {
-  did: string
-  verifiedRoles: Role[]
-}
-
 function App() {
   const [roles, setRoles] = useState<Role[]>([])
+  const [did, setDID] = useState<string>('')
+  const [isLoading, setIsLoading] = useState<Boolean>(false)
 
-  const handleLogin = async () => {
-    const { did } = await iam.initializeConnection({useMetamaskExtension: false})
-    if (!did) {
-      console.log("unable to retrieve DID")
-      return
-    }
-    const signer = iam.getSigner()
-    const latestBlock = await signer?.provider?.getBlockNumber()
+  const verifyIdentity = async function () {
+    const signer = iam.getSigner();
+    const latestBlock = await signer?.provider?.getBlockNumber();
     const claim = await iam.createPublicClaim({
-      data: { blockNumber: latestBlock },
-    })
-    const { data } = await axios.post<{ token: string }>('/login', {
-      claim,
-    })
+      data: { blockNumber: latestBlock }
+    });
+    const { data } = await axios.post<{ token: string }>(
+      "https://did-auth-demo.energyweb.org/login",
+      {
+        claim
+      }
+    );
     const config = {
-      headers: { Authorization: `Bearer ${data.token}` },
+      headers: { Authorization: `Bearer ${data.token}` }
+    };
+    const { data: roles } = await axios.get<Role[]>(
+      "https://did-auth-demo.energyweb.org/roles",
+      config
+    );
+    setRoles(roles);
+  }
+
+  const login = async function (initOptions?: {
+    useMetamaskExtension: boolean;
+  }) {
+    setIsLoading(true);
+    const { did } = await iam.initializeConnection(initOptions);
+    if (did) {
+      setDID(did)
+      try {
+        await verifyIdentity();
+      } catch (err) {
+        console.error(err)
+      }
     }
-    const { data: roles } = await axios.get<Role[]>('/roles', config)
-    const { data: user } = await axios.get<User>('/user', config)
+    setIsLoading(false);
+  }
 
-    setRoles(roles)
+  const loginWithWalletConnect = async function () {
+    setIsLoading(true);
+    const { did } = await iam.initializeConnection();
+    if (did) {
+      setDID(did)
+      try {
+        await verifyIdentity();
+      } catch (err) {
+        console.error(err)
+      }
+    }
+    setIsLoading(false);
   }
-  return (
-    <div className="App">
-      <header className="App-header">
+
+  const logout = function () {
+    setDID('');
+  }
+
+  let loginJsx
```

Comment:
I would assign each view to different function like renderLogin, renderLoading, renderRoles, and main function with
`
if () {
return renderLogin()
}
if () {
return renderButton()
}
return renderRoles()
`
and run this function inside jsx

Reply:
Good point, it's better to not be mutating the `loginJsx` variable